src/pools/ID3MPool.sol

```diff
@@ -0,0 +1,104 @@
+    */
+    function deposit(uint256 amt) external;
+
+    /**
+        @notice Withdraw assets (Dai) from the external pool.
+        @dev If the external pool requires a different amount to be passed in
+        the conversion should occur here as the Hub passes Dai [wad] amounts.
+        msg.sender must be authorized.
+        @param amt amount in asset (Dai) terms that we want to withdraw
+    */
+    function withdraw(uint256 amt) external;
+
+     /**
+        @notice Transfer shares.
+        @dev If the external pool/shares contract requires a different amount to be
+        passed in the conversion should occur here as the Hub passes Gem [wad]
+        amounts. msg.sender must be authorized.
+        @param dst address that should receive the shares
+        @param amt amount in Gem terms that we want to withdraw
+        @return bool whether the transfer was successful per ERC-20 standard
+    */
+    function transfer(address dst, uint256 amt) external returns (bool);
+
+    /**
+        @notice Transfer all shares from this pool.
+        @dev msg.sender must be authorized.
+        @param dst address that should receive the shares.
+        @return bool whether the transfer was successful per ERC-20 standard
+    */
+    function transferAll(address dst) external returns (bool);
+
+    /// @notice Some external pools require manually accruing fees/interest
+    function accrueIfNeeded() external;
+
+    /**
+        @notice Balance of assets this pool "owns".
+        @dev This could be greater than the amount the pool can withdraw due to
+        lack of liquidity.
+        @return uint256 number of assets in Dai [wad]
+    */
+    function assetBalance() external view returns (uint256);
+
+    /**
+        @notice Maximum number of assets the pool could deposit at present.
+        @return uint256 number of assets in Dai [wad]
+    */
+    function maxDeposit() external view returns (uint256);
+
+    /**
+        @notice Maximum number of assets the pool could withdraw at present.
+        @return uint256 number of assets in Dai [wad]
+    */
+    function maxWithdraw() external view returns (uint256);
+
+    /**
+        @notice Used to recover any ERC-20 accidentally sent to the pool.
+        --- YOU SHOULD NOT SEND ERC-20 DIRECTLY TO THIS CONTRACT ---
+        The presence of this function does not convey any right to recover tokens
+        sent to this contract. Maker Governance must evaluate and perform this
+        action at its sole discretion.
+        @dev msg.sender must be authorized.
+        @param token address of the ERC-20 token that will be transferred
+        @param dst address that should receive the shares
+        @param amt amount in Gem terms that we want to withdraw
+        @return bool whether the transfer was successful per ERC-20 standard
+    */
+    function recoverTokens(address token, address dst, uint256 amt) external returns (bool);
```

Comment:
I don't think this needs to be in the interface. Can be added as needed per pool type.

Reply:
this not only protect from sending pool shares where those exist (i.e. cDAI, aDAI), but also if someone sends DAI to the contract. I'm not sure I see the need to have a pool that cannot recover tokens accidentally sent to it.

Reply:
Strictly speaking, shouldn't it prevent "recovering" any of the underlying tokens it uses, specifically pool shares? Otherwise, that's a rug vector (even if it's an accidental rug).

Reply:
Removing from the interface doesn't preclude having a policy of always adding it for safety. Interface design should be minimal to what the API is though and this function is not required to operate the pool.

Reply:
@kmbarry1 I might have got you wrong but I think you are confusing this with a sweep-like functionality that would allow recovering any non related token. The actual intent here is to be able to recover dai sent to the pool directly (as the other functions do not support it). It's correct that any other token can be recovered as well. The potential for confusion is because the function is not authed in the interface, but we do expect any implementation to be authed.

@hexonaut not a stroing opinion, buy I personally see no harm in keeping that in the interface as @iamchrissmith suggested as a way to inforce our policy (it also helps as we have a shared testing infrastructure based on the interface functions).

Reply:
We do tend to copy + paste things like `auth` without requiring an explicit definition in an interface. If you guys all want to keep it, I won't die on this hill, but interfaces should be the minimal set to operate the contract just as a general design practise. These recovery functions are not required to operate the pool.

Reply:
@talbaneth I am not a fan that governance can take the pool shares without respecting the other accounting details--can put a D3M into a totally broken state.

Reply:
lol, not getting into spanking history but aren't we sorry for that decision :)?

I tend to think such a complex system should have recovery abilities not only for mistakenly sent dai, but mostly for unknown bugs/states. So I would say that if we don't make this contract upgradeable we should at least have this extracting option. As @kmbarry1 suggests, we can probably limit it to dai (and not shares) as we have the `quit` function for those (which also handles accounting).

Reply:
Per the discussion on Thursday we decided it's best to just remove it entirely.

Reply:
gone f328965